base_layer/core/src/validation/block_validators.rs

```diff
@@ -118,166 +212,64 @@ impl<B: BlockchainBackend> StatefulValidation<Block, B> for FullConsensusValidat
-
-        Ok(())
-    }
-}
-
-//-------------------------------------     Block validator helper functions     -------------------------------------//
-fn check_accounting_balance(
-    block: &Block,
-    rules: ConsensusManager,
-    factories: &CryptoFactories,
-) -> Result<(), ValidationError>
-{
-    if block.header.height == 0 {
-        // Gen block does not need to be checked for this.
-        return Ok(());
-    }
-    let offset = &block.header.total_kernel_offset;
-    let total_coinbase = rules.calculate_coinbase_and_fees(block);
-    block
-        .body
-        .validate_internal_consistency(&offset, total_coinbase, factories)
-        .map_err(|err| {
-            warn!(
-                target: LOG_TARGET,
-                "Internal validation failed on block:{}:{}",
-                block.hash().to_hex(),
-                err
-            );
-            ValidationError::TransactionError(err)
-        })
-}
-
-fn check_block_weight(block: &Block, consensus_constants: &ConsensusConstants) -> Result<(), ValidationError> {
-    // The genesis block has a larger weight than other blocks may have so we have to exclude it here
-    let block_weight = block.body.calculate_weight();
-    if block_weight <= consensus_constants.get_max_block_transaction_weight() || block.header.height == 0 {
-        trace!(
-            target: LOG_TARGET,
-            "SV - Block contents for block #{} : inputs {}; kernels {}; outputs {}; weight {}.",
-            block.header.height,
-            block.body.inputs().len(),
-            block.body.kernels().len(),
-            block.body.outputs().len(),
-            block_weight,
-        );
-        Ok(())
-    } else {
-        Err(BlockValidationError::BlockTooLarge).map_err(ValidationError::from)
-    }
-}
-
-fn check_coinbase_output(
-    block: &Block,
-    rules: ConsensusManager,
-    factories: &CryptoFactories,
-) -> Result<(), ValidationError>
-{
-    trace!(
-        target: LOG_TARGET,
-        "Checking coinbase output on block with hash {}",
-        block.hash().to_hex()
-    );
-    let total_coinbase = rules.calculate_coinbase_and_fees(block);
-    block
-        .check_coinbase_output(
-            total_coinbase,
-            rules.consensus_constants(block.header.height),
-            factories,
-        )
-        .map_err(ValidationError::from)
-}
-
 // This function checks for duplicate inputs and outputs. There should be no duplicate inputs or outputs in a block
-fn check_duplicate_transactions_inputs(block: &Block) -> Result<(), ValidationError> {
-    trace!(
-        target: LOG_TARGET,
-        "Checking duplicate inputs and outputs on block with hash {}",
-        block.hash().to_hex()
-    );
-    for i in 1..block.body.inputs().len() {
-        if block.body.inputs()[i..].contains(&block.body.inputs()[i - 1]) {
+fn check_duplicate_transactions_inputs(body: &AggregateBody) -> Result<(), ValidationError> {
+    for i in 1..body.inputs().len() {
```

Comment:
If the inputs are sorted, can't you just do one pass and see if it matches the previous input?

Reply:
You're right, and I think we must sort the inputs and outputs before validating even in initial block sync (we can't rely that the MR in the header is valid). Looking at Vec's built-in sort implementation, it is very efficient at sorting Vecs that are already mostly sorted, so I think this is a great optimisation.